samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -29,97 +28,116 @@
 import org.apache.samza.operators.windows.WindowPane;
 
 import java.util.Collection;
+import java.util.function.Function;
 
 
 /**
- * Represents a stream of {@link MessageEnvelope}s.
+ * Represents a stream of messages.
  * <p>
  * A {@link MessageStream} can be transformed into another {@link MessageStream} by applying the transforms in this API.
  *
- * @param <M>  type of {@link MessageEnvelope}s in this stream
+ * @param <M>  type of messages in this stream
  */
 @InterfaceStability.Unstable
-public interface MessageStream<M extends MessageEnvelope> {
+public interface MessageStream<M> {
 
   /**
-   * Applies the provided 1:1 {@link MapFunction} to {@link MessageEnvelope}s in this {@link MessageStream} and returns the
+   * Applies the provided 1:1 function to messages in this {@link MessageStream} and returns the
    * transformed {@link MessageStream}.
    *
-   * @param mapFn the function to transform a {@link MessageEnvelope} to another {@link MessageEnvelope}
-   * @param <TM> the type of {@link MessageEnvelope}s in the transformed {@link MessageStream}
+   * @param mapFn the function to transform a message to another message
+   * @param <TM> the type of messages in the transformed {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM extends MessageEnvelope> MessageStream<TM> map(MapFunction<M, TM> mapFn);
+  <TM> MessageStream<TM> map(MapFunction<M, TM> mapFn);
```

Comment:
Agree that filter/map/flatMap can be implemented in terms of Java's Function, but I wanted to add separate classes for our functions in case we need to add additional behavior to them (e.g. serializability, initialization, access to a shared context in apply etc).